keps/sig-api-machinery/20190723-coordinated-discovery-document.md

```diff
@@ -0,0 +1,181 @@
+---
+title: Coordinated Discovery Documents for HA API Servers
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@yliaog"
+  - "@lavalamp"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2019-07-23
+last-updated: 2019-07-25
+status: provisional
+---
+
+# Coordinated Discovery Documents for HA API Servers
+
+## Table of Contents
+
+* [Problem Statement](#problem-statement)
+* [Summary of Proposed changes](#summary-of-proposed-changes)
+* [API changes](#api-changes)
+* [Changes to API servers](#changes-to-api-servers)
+   * [Calculating DiscoveryDocHashes](#calculating-discoverydochashes)
+   * [Periodically reporting DiscoveryDocHashes &amp; Garbage Collection](#periodically-reporting-discoverydochashes--garbage-collection)
+   * [Handling discovery requests](#handling-discovery-requests)
+* [Backwards Compatibility](#backwards-compatibility)
+* [Graduation Plan](#graduation-plan)
+* [Alternatives](#alternatives)
+
+## Problem Statement
+
+API servers today compute discovery documents based on their local states. When
+HA API servers are rolling upgraded, API server instances can return
+different discovery documents to clients. This causes problems. For
+example, the storage migrator does not work reliably with HA API servers
+([details][]).
+
+[details]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md#ha-masters
+
+This KEP proposes a mechanism for HA API servers to expose to clients if
+all API server instances are serving consistent discovery document.
+
+## Summary of Proposed changes
+
+Periodically, each API server instance in an HA setup calculates a hash based on
+the discovery documents of the built-in APIs. The hash value is reported in a
+new API, `DiscoveryDocHashes`, together with the API server's ID. 
+
+When responding to discovery document requests, apart from the original
+discovery documents, the API server also sets a `Consistent` field to indicate if
+all API servers are serving the same discovery documents, i.e., if all hash
+values recorded in `DiscoveryDocHashes` are equal. 
+
+## API changes
+
+We introduce a new API `DiscoveryDocHashes`, in a new API group
+`APIServerCoordination.k8s.io`. 
+
+```golang
+type DiscoveryDocHashes struct {
+   Hashes []Hash
+}
+
+type Hash struct {
+   // The ID of the reporting API server. 
+   // The ID can be generated in a similar fashion as the controller manager
+   // generates leader election id at
+   // https://github.com/kubernetes/kubernetes/blob/2321d1e9e8950cc94b3ef2368dfaacec61f1ba4f/cmd/cloud-controller-manager/app/controllermanager.go#L185.
+   // Required.
+   APIServerID string
+
+   // A hash calculated based on the discovery documents of built-in APIs. If
+   // two hashes equal, it means that the two API servers serve the same set of
+   // APIs.
+   // Required.
+   BuiltInAPIHash string
```

Comment:
What about CRDs? Aggregated APIs?

I think this system can be broken by an aggregated apiserver which returns a discovery doc that isn't constant (e.g., maybe a documentation string changes for some reason).